src/core/api.c

```diff
@@ -1125,12 +1120,29 @@ MsQuicStreamSend(
     } else if (QueueOper) {
         Oper = QuicOperationAlloc(Connection->Worker, QUIC_OPER_TYPE_API_CALL);
         if (Oper == NULL) {
-            Status = QUIC_STATUS_OUT_OF_MEMORY;
             QuicTraceEvent(
                 AllocFailure,
                 "Allocation of '%s' failed. (%llu bytes)",
                 "STRM_SEND operation",
                 0);
+            //
+            // We can't fail the send at this point, because we're already queued
+            // the send above. So instead, we're just going to abort the whole
+            // connection.
+            //
+            if (InterlockedCompareExchange16(
+                    (short*)&Connection->BackUpOperUsed, 1, 0) != 0) {
+                goto Exit; // It's already started the shutdown.
+            }
+            Oper = &Connection->BackUpOper;
+            Oper->FreeAfterProcess = FALSE;
+            Oper->Type = QUIC_OPER_TYPE_API_CALL;
+            Oper->API_CALL.Context = &Connection->BackupApiContext;
+            Oper->API_CALL.Context->Type = QUIC_API_TYPE_CONN_SHUTDOWN;
+            Oper->API_CALL.Context->CONN_SHUTDOWN.Flags = QUIC_CONNECTION_SHUTDOWN_FLAG_SILENT;
+            Oper->API_CALL.Context->CONN_SHUTDOWN.ErrorCode = 0;
```

Comment:
Will the error code be sent to the peer? I wonder if we should fill in the OOM status code so the app knows what killed it.

Reply:
Currently, with how this is written, the app just gets a shutdown complete event, no error code. It would take a bigger change to fix that. I'll look into it tomorrow.

Reply:
Yeah, not ideal. I will have to refactor some stuff to make this work better.

Reply:
Fixed.